SPANNER this.spannerSchemaUtils.getKey(entity) is failing wiht Nullpointerexception

This is happening when one of the key of hte composite primary key is null. It used to work perfectly till afternoon.



```
com.google.cloud.spanner.SpannerException: UNKNOWN
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerExceptionPreformatted(SpannerExceptionFactory.java:119)
	at com.google.cloud.spanner.SpannerExceptionFactory.newSpannerException(SpannerExceptionFactory.java:43)
	at com.google.cloud.spanner.SpannerImpl$TransactionRunnerImpl.runInternal(SpannerImpl.java:1313)
	at com.google.cloud.spanner.SpannerImpl$TransactionRunnerImpl.run(SpannerImpl.java:1266)
	at com.google.cloud.spanner.SessionPool$PooledSession$1.run(SessionPool.java:398)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.lambda$performReadWriteTransaction$8(SpannerTemplate.java:307)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.doWithOrWithoutTransactionContext(SpannerTemplate.java:495)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate.performReadWriteTransaction(SpannerTemplate.java:303)
	at com.homedepot.loyalty.dao.LoyaltyBaseDAO.saveOrUpdate(LoyaltyBaseDAO.java:80)
	at com.homedepot.loyalty.repository.impl.ProgramRepositoryImpl.addProducts(ProgramRepositoryImpl.java:243)
	at com.homedepot.loyalty.repository.impl.ProgramRepositoryImpl$$FastClassBySpringCGLIB$$bcfc7556.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:746)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:185)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
	at com.homedepot.loyalty.repository.impl.ProgramRepositoryImpl$$EnhancerBySpringCGLIB$$9e0e8d16.addProducts(<generated>)
	at com.homedepot.loyalty.service.ProgramAddProductService.addProducts(ProgramAddProductService.java:30)
	at com.homedepot.loyalty.controller.ProgramController.addProductsToProgram(ProgramController.java:161)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:209)
	at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:136)
	at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:102)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:877)
	at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:783)
	at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:991)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:925)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:974)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:848)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:742)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.filterAndRecordMetrics(WebMvcMetricsFilter.java:158)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.filterAndRecordMetrics(WebMvcMetricsFilter.java:126)
	at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:111)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.boot.actuate.web.trace.servlet.HttpTraceFilter.doFilterInternal(HttpTraceFilter.java:90)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HttpPutFormContentFilter.doFilterInternal(HttpPutFormContentFilter.java:109)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:198)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:496)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:140)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:87)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:650)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:342)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:803)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:790)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1468)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.springframework.cloud.gcp.data.spanner.core.convert.ConverterAwareMappingSpannerEntityWriter.convertKeyPart(ConverterAwareMappingSpannerEntityWriter.java:191)
	at org.springframework.cloud.gcp.data.spanner.core.convert.ConverterAwareMappingSpannerEntityWriter.convertToKey(ConverterAwareMappingSpannerEntityWriter.java:169)
	at org.springframework.cloud.gcp.data.spanner.core.convert.ConverterAwareMappingSpannerEntityProcessor.convertToKey(ConverterAwareMappingSpannerEntityProcessor.java:192)
	at org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerCompositeKeyProperty.getId(SpannerCompositeKeyProperty.java:87)
	at org.springframework.cloud.gcp.data.spanner.core.mapping.SpannerPersistentEntityImpl$1.getProperty(SpannerPersistentEntityImpl.java:394)
	at org.springframework.cloud.gcp.data.spanner.core.admin.SpannerSchemaUtils.getKey(SpannerSchemaUtils.java:98)
	at com.homedepot.loyalty.dao.LoyaltyBaseDAO.saveOrUpdate(LoyaltyBaseDAO.java:96)
	at com.homedepot.loyalty.dao.LoyaltyBaseDAO.lambda$1(LoyaltyBaseDAO.java:81)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at com.homedepot.loyalty.dao.LoyaltyBaseDAO.lambda$0(LoyaltyBaseDAO.java:81)
	at org.springframework.cloud.gcp.data.spanner.core.SpannerTemplate$1.run(SpannerTemplate.java:320)
	at com.google.cloud.spanner.SpannerImpl$TransactionRunnerImpl.runInternal(SpannerImpl.java:1298)
	... 80 more
```

--- a/spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java
+++ b/spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/convert/ConverterAwareMappingSpannerEntityWriter.java
@@ -188,7 +188,7 @@ public class ConverterAwareMappingSpannerEntityWriter implements SpannerEntityWr
 
 	private Object convertKeyPart(Object object) {
 
-		if (isValidSpannerKeyType(ConversionUtils.boxIfNeeded(object.getClass()))) {
+		if (object == null || isValidSpannerKeyType(ConversionUtils.boxIfNeeded(object.getClass()))) {
 			return object;
 		}
 		/*

--- a/spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerCompositeKeyProperty.java
+++ b/spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerCompositeKeyProperty.java
@@ -65,23 +65,24 @@ public class SpannerCompositeKeyProperty implements SpannerPersistentProperty {
 		PersistentPropertyAccessor accessor = getOwner().getPropertyAccessor(entity);
 		List keyParts = new ArrayList();
 		for (SpannerPersistentProperty spannerPersistentProperty : this.primaryKeyColumns) {
+			Object value = accessor.getProperty(spannerPersistentProperty);
 			if (spannerPersistentProperty.isEmbedded()) {
 				Key embeddedKeyParts = this.spannerPersistentEntity
 						.getSpannerMappingContext()
 						.getPersistentEntity(spannerPersistentProperty.getType())
 						.getIdProperty()
-						.getId(accessor.getProperty(spannerPersistentProperty));
+						.getId(value);
 				for (Object keyPart : embeddedKeyParts.getParts()) {
 					keyParts.add(keyPart);
 				}
 			}
-			else if (spannerPersistentProperty.getAnnotatedColumnItemType() != null) {
-				keyParts.add(this.spannerPersistentEntity.getSpannerEntityWriter().getSpannerWriteConverter()
-						.convert(accessor.getProperty(spannerPersistentProperty), SpannerTypeMapper
-								.getSimpleJavaClassFor(spannerPersistentProperty.getAnnotatedColumnItemType())));
+			else if (spannerPersistentProperty.getAnnotatedColumnItemType() == null || value == null) {
+				keyParts.add(value);
 			}
 			else {
-				keyParts.add(accessor.getProperty(spannerPersistentProperty));
+				keyParts.add(this.spannerPersistentEntity.getSpannerEntityWriter().getSpannerWriteConverter()
+						.convert(value, SpannerTypeMapper
+								.getSimpleJavaClassFor(spannerPersistentProperty.getAnnotatedColumnItemType())));
 			}
 		}
 		return this.spannerPersistentEntity.getSpannerEntityWriter().convertToKey(keyParts);

--- a/spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerPersistentEntityImplTests.java
+++ b/spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerPersistentEntityImplTests.java
@@ -234,11 +234,14 @@ public class SpannerPersistentEntityImplTests {
 		childEmbedded.parentEmbedded = parentEmbedded;
 		childEmbedded.id4 = "4";
 
+		// intentionally null, which is a supported key component type.
+		childEmbedded.id5 = null;
+
 		Key key = (Key) this.spannerMappingContext
 				.getPersistentEntity(ChildEmbedded.class)
 				.getIdentifierAccessor(childEmbedded).getIdentifier();
 		assertEquals(
-				Key.newBuilder().append("1").append("2").append("3").append("4").build(),
+				Key.newBuilder().append("1").append("2").append("3").append("4").appendObject(null).build(),
 				key);
 	}
 
@@ -258,7 +261,7 @@ public class SpannerPersistentEntityImplTests {
 	@Test
 	public void testExcludeEmbeddedColumnNames() {
 		assertThat(this.spannerMappingContext.getPersistentEntity(ChildEmbedded.class)
-				.columns(), containsInAnyOrder("id", "id2", "id3", "id4"));
+				.columns(), containsInAnyOrder("id", "id2", "id3", "id4", "id5"));
 	}
 
 	@Test
@@ -356,6 +359,10 @@ public class SpannerPersistentEntityImplTests {
 
 		@PrimaryKey(keyOrder = 2)
 		String id4;
+
+		@PrimaryKey(keyOrder = 3)
+		@Column(spannerType = TypeCode.STRING)
+		Long id5;
 	}
 
 	private static class ChildCollectionEmbedded {

